File: NuwaKext/SocketFilter/DNSResolver.cpp
//
//  DNSResolver.cpp
//  NuwaKext
//
//  Created by ConradSun on 2022/8/30.
//

#include "DNSResolver.hpp"
#include "KextLogger.hpp"

#pragma mark - DNS Resolver

DNSResolver::DNSResolver(const char *packet, size_t size, UInt8 proto) {
    m_originPacket = packet;
    m_packetSize = size;
    m_protocol = proto;
    m_parseResults = {0, nullptr};
    bzero(&m_domainMap, sizeof(m_domainMap));
}

DNSResolver::~DNSResolver() {
    if (m_parseResults.results != nullptr) {
        IOFreeAligned(m_parseResults.results, sizeof(DNSParseResult)*m_parseResults.count);
        m_parseResults.results = nullptr;
    }
    
    for (UInt8 i = 0; i < kMaxNameCount; ++i) {
        if (m_domainMap[i].domainName == nullptr) {
            break;
        }
        size_t size = strlen(m_domainMap[i].domainName) + 1;
        IOFreeAligned(m_domainMap[i].domainName, size);
        m_domainMap[i].domainName = nullptr;
    }
}

bool DNSResolver::setDoaminMap(const char *domainName, UInt16 index) {
    DNSDomainMap *map = nullptr;
    for (UInt8 i = 0; i < kMaxNameCount; ++i) {
        if (m_domainMap[i].domainName == nullptr) {
            map = &m_domainMap[i];
            break;
        }
    }
    if (map == nullptr) {
        return false;
    }
    
    size_t size = strlen(domainName) + 1;
    map->domainName = (char *)IOMallocAligned(size, 2);
    if (map->domainName == nullptr) {
        return false;
    }
    
    strlcpy(map->domainName, domainName, size);
    map->index = index;
    return true;
}

SInt16 DNSResolver::getDomainIndex(const char *domainName) {
    for (UInt8 i = 0; i < kMaxNameCount; ++i) {
        if (m_domainMap[i].domainName == nullptr) {
            return -1;
        }
        if (strcmp(domainName, m_domainMap[i].domainName) == 0) {
            return m_domainMap[i].index;
        }
    }
    return -1;
}

bool DNSResolver::parseDomainName(const char *nameBegin, char *domainName, UInt16 nameSize) {
    if (nameBegin == nullptr || domainName == nullptr || nameSize == 0) {
        return false;
    }
    
    static char offsetSymbol = 0xc0;
    static char endSymbol = 0x00;
    const char *parseBegin = nameBegin;
    const char *parseEnd = m_originPacket + m_packetSize;
    
    UInt16 nameLen = 0;
    UInt16 occupyCount = 0;
    
    if (*parseBegin == offsetSymbol) {
        occupyCount = 2;
        parseBegin = m_originPacket + *(parseBegin + 1);
        Logger(LOG_DEBUG, "Domain address is offseted.")
    }
    
    while (parseBegin < parseEnd && *parseBegin != endSymbol) {
        if (*parseBegin == offsetSymbol) {
            domainName[nameLen++] = '.';
            occupyCount = occupyCount == 0 ? (parseBegin - nameBegin + 1) : occupyCount;
            m_parseIndex += occupyCount;
            return parseDomainName(parseBegin, domainName+nameLen, nameSize-nameLen);
        }
        
        UInt8 count = *parseBegin++;
        if (nameLen != 0) {
            domainName[nameLen++] = '.';
        }
        if (count >= (nameSize-nameLen)) {
            Logger(LOG_ERROR, "Domain name is too long.")
            return false;
        }
        if (count >= (parseEnd-parseBegin)) {
            Logger(LOG_ERROR, "Domain name is invalid.")
            return false;
        }
        for (UInt8 i = 0; i < count; ++i) {
            domainName[nameLen++] = *parseBegin++;
        }
    }
    domainName[nameLen] = '\0';
    
    occupyCount = occupyCount == 0 ? (parseBegin - nameBegin + 1) : occupyCount;
    m_parseIndex += occupyCount;
    return true;
}

bool DNSResolver::parseQuery(UInt16 queryCount, UInt16 replyCode) {
    for (UInt16 i = 0; i < queryCount; ++i) {
        m_parseResults.results[i].replyCode = replyCode;
        if (!parseDomainName(m_originPacket+m_parseIndex, m_parseResults.results[i].domainName, kMaxNameLength)) {
            return false;
        }
        
        m_parseIndex += kDNSQuerySize;
        if (m_parseIndex >= m_packetSize) {
            return false;
        }
        if (!setDoaminMap(m_parseResults.results[i].domainName, i)) {
            return false;
        }
    }
    
    return true;
}

bool DNSResolver::processReplyResult(const char *domain, const char *result, UInt16 type) {
    SInt16 index = getDomainIndex(domain);
    if (index < 0) {
        Logger(LOG_ERROR, "Unknown reply domain name [%s].", result)
        return false;
    }
    
    if (type == kDNSType_CNAME) {
        return setDoaminMap(result, index);
    }
    else {
        size_t length = strlen(m_parseResults.results[index].queryResult);
        if (length + strlen(result) + 1 >= kMaxPathLength) {
            Logger(LOG_ERROR, "Reply info is too much.")
            return false;
        }
        if (length > 0) {
            m_parseResults.results[index].queryResult[length++] = ',';
        }
        strlcpy(m_parseResults.results[index].queryResult+length, result, kMaxPathLength-length);
    }
    
    return true;
}

bool DNSResolver::parseReplyItem(DNSParseResult *result) {
    if (!parseDomainName(m_originPacket+m_parseIndex, result->domainName, kMaxNameLength)) {
        return false;
    }
    
    DNSResponseInfo info = *(DNSResponseInfo *)(m_originPacket + m_parseIndex);
    info.DNSType = ntohs(info.DNSType);
    info.DNSClass = ntohs(info.DNSClass);
    info.liveTime = ntohl(info.liveTime);
    info.length = ntohs(info.length);
    result->replyType = info.DNSType;
    
    const char *messageBegin = m_originPacket + m_parseIndex + kDNSReplySize;
    m_parseIndex += info.length;
    switch (result->replyType) {
        case kDNSType_A:
            inet_ntop(AF_INET, messageBegin, result->queryResult, kMaxPathLength);
            Logger(LOG_DEBUG, "The replied result is IPv4 [%s].", result->queryResult)
            break;
        case kDNSType_CNAME:
            parseDomainName(messageBegin, result->queryResult, kMaxPathLength);
            m_parseIndex -= info.length;
            Logger(LOG_DEBUG, "The replied result is canonical name [%s].", result->queryResult)
            break;
        case kDNSType_AAAA:
            inet_ntop(AF_INET6, messageBegin, result->queryResult, kMaxPathLength);
            Logger(LOG_DEBUG, "The replied result is IPv6 [%s].", result->queryResult)
            break;
        default:
            return true;
    }
    
    return processReplyResult(result->domainName, result->queryResult, result->replyType);
}

bool DNSResolver::parseReply(UInt16 replyCount) {
    DNSParseResult tempResult = {};
    for (UInt16 i = 0; i < replyCount; ++i) {
        bzero(&tempResult, sizeof(DNSParseResult));
        if (!parseReplyItem(&tempResult)) {
            return false;
        }
        
        m_parseIndex += kDNSReplySize;
        if (m_parseIndex > m_packetSize) {
            return false;
        }
    }
    
    return true;
}

void DNSResolver::parsePacket() {
    DNSMessageHeader header = *(DNSMessageHeader *)m_originPacket;
    header.transID = ntohs(header.transID);
    header.flags = ntohs(header.flags);
    header.questions = ntohs(header.questions);
    header.answers = ntohs(header.answers);
    
    UInt16 replyCode = header.flags & 0x000f;
    size_t allocSize = sizeof(DNSParseResult) * header.questions;
    m_parseResults.results = (DNSParseResult *)IOMallocAligned(allocSize, 2);
    if (m_parseResults.results == nullptr) {
        return;
    }
    m_parseResults.count = header.questions;
    bzero(m_parseResults.results, allocSize);
    
    m_parseIndex += kDNSHeaderSize;
    if (!parseQuery(header.questions, replyCode) || !parseReply(header.answers)) {
        IOFreeAligned(m_parseResults.results, allocSize);
        m_parseResults.count = 0;
        m_parseResults.results = nullptr;
        return;
    }
}

DNSResolveResults DNSResolver::getResults() {
    if (m_originPacket == nullptr || m_packetSize == 0) {
        return m_parseResults;
    }
    if (m_protocol == IPPROTO_UDP) {
        if (m_packetSize <= kDNSHeaderSize) {
            return m_parseResults;
        }
        m_parseIndex = 0;
    }
    else if (m_protocol == IPPROTO_TCP) {
        if (m_packetSize <= sizeof(UInt16) + kDNSHeaderSize) {
            return m_parseResults;
        }
        // skip the length field of the header
        m_parseIndex = sizeof(UInt16);
    }
    else {
        return m_parseResults;
    }
    
    parsePacket();
    return m_parseResults;
}

File: NuwaSext/NetworkExtension/DNSResolver.swift
//
//  DNSResolver.swift
//  NuwaSext
//
//  Created by ConradSun on 2022/9/6.
//

import Foundation
import Network

fileprivate enum TypeCode: UInt {
    case A      = 1
    case CNAME  = 5
    case AAAA   = 28
}

fileprivate enum ClassCode: UInt {
    case IN     = 1
    case S      = 2
    case CH     = 3
    case HS     = 4
    case NONE   = 254
    case ANY    = 255
}

fileprivate struct MessageHeader {
    var transID: UInt16
    var flags: UInt16
    var questions: UInt16
    var answers: UInt16
    var authorities: UInt16
    var additionals: UInt16
    init() {
        transID = 0
        flags = 0
        questions = 0
        answers = 0
        authorities = 0
        additionals = 0
    }
}

fileprivate struct QueryInfo {
    var dnsType: UInt16
    var dnsClass: UInt16
}

fileprivate struct ResponseInfo {
    var dnsType: UInt16
    var dnsClass: UInt16
    var liveTime: UInt32
    var length: UInt16
    init() {
        dnsType = 0
        dnsClass = 0
        liveTime = 0
        length = 0
    }
}

struct DNSParseResult {
    var replyCode: UInt16
    var domainName: String
    var queryResult: String
    init() {
        replyCode = 0
        domainName = ""
        queryResult = ""
    }
}

class DNSResolver {
    private let headerSize = 12
    private let querySize = 4
    private let replySize = 10
    
    private var originData = Data()
    private var parseIndex = 0
    private var nameDict = [String: UInt16]()
    var results = [DNSParseResult]()
    
    private func ntohs(data: Data, index: Int) -> UInt16 {
        return (UInt16(data[index]) << 8) | UInt16(data[index+1])
    }
    
    private func parseDomainName(begin: Int, domainName: inout String) ->Bool {
        if begin >= originData.count {
            return false
        }
        
        let offsetSymbol:UInt8 = 0xc0
        let endSymbol:UInt8 = 0x00
        var currentSite = begin
        var occupyCount = 0
        
        if originData[currentSite] == offsetSymbol {
            occupyCount = 2
            currentSite = Int(originData[currentSite + 1])
            Logger(.Debug, "Domain address is offseted.")
        }
        while currentSite < originData.count && originData[currentSite] != endSymbol {
            if originData[currentSite] == offsetSymbol {
                domainName += "."
                occupyCount = occupyCount == 0 ? (currentSite - begin + 1) : occupyCount
                parseIndex += occupyCount
                return parseDomainName(begin: currentSite, domainName: &domainName)
            }
            
            let count = originData[currentSite]
            currentSite += 1
            if !domainName.isEmpty {
                domainName += "."
            }
            if count >= (originData.count-currentSite) {
                Logger(.Error, "Domain name is invalid.")
                return false
            }
            for _ in 0 ..< count {
                let char = Character(UnicodeScalar(originData[currentSite]))
                domainName.append(char)
                currentSite += 1
            }
        }
        
        occupyCount = occupyCount == 0 ? (currentSite - begin + 1) : occupyCount
        parseIndex += occupyCount
        return true
    }
    
    private func parseQuery(count: UInt16, rcode: UInt16) -> Bool {
        for i in 0 ..< count {
            var result = DNSParseResult()
            result.replyCode = rcode
            if !parseDomainName(begin: parseIndex, domainName: &result.domainName) {
                return false
            }
            parseIndex += querySize
            results.append(result)
            nameDict.updateValue(i, forKey: result.domainName)
        }
        return true
    }
    
    private func processIPAddr(addrFamily: Int32, addr: UnsafeRawPointer, domainName: String) {
        guard let index = nameDict[domainName] else {
            return
        }
        
        var tempString = [CChar](repeating: 0, count: Int(MAXPATHLEN))
        inet_ntop(addrFamily, addr, &tempString, socklen_t(MAXPATHLEN))
        
        if !results[Int(index)].queryResult.isEmpty {
            results[Int(index)].queryResult += ","
        }
        results[Int(index)].queryResult += String(cString: &tempString)
    }
    
    private func processCanonicalName(begin: Int, domainName: String) {
        guard let index = nameDict[domainName] else {
            return
        }
        
        var name = ""
        if !parseDomainName(begin: begin, domainName: &name) {
            return
        }
        nameDict.updateValue(index, forKey: name)
    }
    
    private func parseReplyItem() -> Bool {
        var result = DNSParseResult()
        if !parseDomainName(begin: parseIndex, domainName: &result.domainName) {
            return false
        }
        if originData.count - parseIndex <= replySize {
            return false
        }
        
        var info = ResponseInfo()
        info.dnsType = ntohs(data: originData, index: parseIndex)
        info.length = ntohs(data: originData, index: parseIndex+8)
        
        let type = TypeCode(rawValue: UInt(info.dnsType))
        originData.withUnsafeBytes { pointer in
            
            guard var addr = pointer.baseAddress else {
                return
            }
            addr += parseIndex + replySize
            
            switch type {
            case .A:
                processIPAddr(addrFamily: AF_INET, addr: addr, domainName: result.domainName)
            case .CNAME:
                processCanonicalName(begin: parseIndex + replySize, domainName: result.domainName)
                parseIndex -= Int(info.length)
            case .AAAA:
                processIPAddr(addrFamily: AF_INET6, addr: addr, domainName: result.domainName)
            default:
                break
            }
        }
        parseIndex += Int(info.length)
        return true
    }
    
    private func parseReply(count: UInt16) -> Bool {
        for _ in 0 ..< count {
            if !parseReplyItem() {
                return false
            }
            parseIndex += replySize
        }
        return true
    }
    
    func parseMessage(message: Data, proto: Int32) {
        if proto == IPPROTO_UDP {
            if message.count <= headerSize {
                return
            }
            parseIndex = 0
        }
        else if proto == IPPROTO_TCP {
            if message.count <= headerSize + MemoryLayout<UInt16>.size {
                return
            }
            parseIndex = MemoryLayout<UInt16>.size
        }
        else {
            return
        }
        originData = message
        
        var header = MessageHeader()
        header.transID = ntohs(data: originData, index: parseIndex)
        header.flags = ntohs(data: originData, index: parseIndex+2)
        header.questions = ntohs(data: originData, index: parseIndex+4)
        header.answers = ntohs(data: originData, index: parseIndex+6)
        parseIndex += headerSize
        let replyCode = header.flags & 0x000f
        
        if !parseQuery(count: header.questions, rcode: replyCode) || !parseReply(count: header.answers) {
            Logger(.Warning, "Failed to parse dns packet.")
            results.removeAll()
            return
        }
    }
}
